app/background.cc

```diff
@@ -36,62 +37,62 @@ static void read_thread(std::shared_ptr<background_t> pbkd) {
         if (pbkd->cap.read(grab)) {
             // got a frame, lock and copy for callers
             {
-                std::unique_lock<std::mutex> hold(pbkd->mux);
+                std::unique_lock<std::mutex> hold(pbkd->rawmux);
                 grab.copyTo(pbkd->raw);
-                pbkd->frm += 1;
+                pbkd->frame += 1;
             }
             // grab timing point
             auto now = std::chrono::steady_clock::now();
-            // display thumbnail frame with overlay info if double debug enabled
-            if (pbkd->dbg > 1) {
+            // generate thumbnail frame with overlay info if double debug enabled
+            if (pbkd->debug > 1) {
                 char msg[40];
                 long nsec = std::chrono::duration_cast<std::chrono::nanoseconds>(now-last).count();
                 double fps = 1e9/(double)nsec;
-                // allow a few settling frames..
-                if (pbkd->frm > 10) {
-                    min = (0.0==min || fps<min) ? fps : min;
-                    max = (0.0==max || fps>max) ? fps : max;
+                {
+                    std::unique_lock<std::mutex> hold(pbkd->thumbmux);
+                    cv::resize(grab, pbkd->thumb, cv::Size(160,120));
+                    snprintf(msg, sizeof(msg), "FPS:%0.1f", fps);
+                    cv::putText(pbkd->thumb, msg, cv::Point(5,15), cv::FONT_HERSHEY_PLAIN, 1.0, cv::Scalar(0,255,255));
+                    snprintf(msg, sizeof(msg), "FRM:%05d", fps, pbkd->frame);
+                    cv::putText(pbkd->thumb, msg, cv::Point(5,30), cv::FONT_HERSHEY_PLAIN, 1.0, cv::Scalar(0,255,255));
+                    cv::putText(pbkd->thumb, "Background", cv::Point(5,pbkd->thumb.rows-5), cv::FONT_HERSHEY_PLAIN, 1.0, cv::Scalar(0,255,255));
```

Comment:
Minimize work done under active lock. Use thread-local buffer to prepare and only lock once the thumbnail is ready.

Second thing: the thumbnail should keep the ratio from the input/output frame; maybe 1:8?

Reply:
good point, copying `cv::Mat` objects is low cost (header only), so working in a local and copying out would be fine.